### DISCUSSION TITLE
VSA 2020 QGIS project configuration

### DISCUSSION THREAD

**Discussion by ponceta** [QGEP/QGEP#828] [Ideas]
According to the future layers change which will happen with VSA DSS 2020 implementations, here are some insights of what we want for the future QGIS project (knowing that QGIS has evolved much between the first implementations) (initial proposition from @urskaufmann and following discussions of the technical group https://github.com/teksi/Home/wiki/Meeting-2023-08-08-(Technical-Group) )

**New standards for Forms configuration**

 - [ ] Standard options for fields with value lists: 
->Null allowed, 
->sorted by values, 
->Description-Field as description,
-> Filter: field active, 
-> no default value (but document how to set it)

![Standard options for fields with valuelist_add_active_filtery](https://github.com/QGEP/QGEP/assets/1265865/0f5d2062-6b3b-4172-86da-bcbe1e985035)
![Standard options for fields with valuelist_mandatory](https://github.com/QGEP/QGEP/assets/1265865/53e06242-aaf9-476d-bde3-7c752e30cd8e)
![Standard options for fields with valuelis_add_description_en](https://github.com/QGEP/QGEP/assets/1265865/aa52780f-1732-434a-9cf1-de7bff6011d4)
![Standard options for fields with valuelist](https://github.com/QGEP/QGEP/assets/1265865/e3d9e9e4-29ec-49fb-bffd-1bc9aa9730fc)

 - [ ] Numeric fields with range in vsa-dss: 
->As range till now (problem qgis corrects without  warning to the limit)
->as range with bigger limits and constraints: best as
->text and constraints, s
->some cases as editable unique values and constraints 

![Standard options for fields numeric option text edit](https://github.com/QGEP/QGEP/assets/1265865/943dda9a-44c7-4d0b-9934-1727191cb03c)
![Standard options for fields numeric option unique values eg  for years](https://github.com/QGEP/QGEP/assets/1265865/615ada09-d4ed-4102-a1a9-1bdd7f1007f5)
![Standard options for fields numeric](https://github.com/QGEP/QGEP/assets/1265865/04d27ad9-063e-45ce-adea-ab4550228a3f)

 - [ ] Obj_id is generally visible, but not editable 
-> (read only)
 
![obj_id_settings](https://github.com/QGEP/QGEP/assets/1265865/1d65d194-c5ef-4929-b51c-cd16c05d3633)
![Standard options for fields identifier_is_obj_id](https://github.com/QGEP/QGEP/assets/1265865/d0ef6d1a-3228-4dc9-85ce-6ee8d78dfb31)

Take out editable:
![obj_id_take_out_editable](https://github.com/QGEP/QGEP/assets/1265865/a316aa4a-d0bf-4dc8-b2e3-95db0e0ed6b9)

 - [ ] Reuse last value 
–> option with most fields on the first tab of the form

![Standard options for fields with valuelist_reuse_last_entered_value_for_general_tab_attributes](https://github.com/QGEP/QGEP/assets/1265865/e9041102-cf35-4e78-8a3c-7c9a2b9b9b85)

 - [ ] Colors are not used now, is there a concept for colors? 
->[see new issue](https://github.com/QGEP/QGEP/issues/827)

 - [ ] (what are the correct constraints / defaults for identifier and obj_id-fields): obj_id: 
 -> four option in constraints checked, identifier: not set in database, but set defaultvalue = obj_id
![rp_from_obj_id_settings](https://github.com/QGEP/QGEP/assets/1265865/a4dcfd9a-fe7e-4c8b-bfd3-1b4d3d80411e)

**Additionnal points :**

 - [ ] Hardcoded solutions replaced by user preferences:
e.g.: which inlets and outlets are labeled at the _labels wastewater structure (today defined in FUNCTION: qgep_od.update_wastewater_structure_label): some user label all input_level, some user label only paa-input_level -> _label_input_paa and _label_input_all.

- [ ] Addition for the views: The whole handling of SAA and PAA is quite tedious because of the combination with the function_hierarchical, one has to enter again and again these code values for PAA in filters, symbolizations etc.. If in vw_qgep_reach a calculated field _paa (resp. _pwwf) would be available and there simply 0 and 1 or true and false or pwwf and swwf would be entered, whether PAA or SAA, then this would simplify some of the recurring work in symbolization, queries etc.
Analog wäre ein gleiches Feld im vw_wastewater_node und im  vw_qgep_wasterwater_structure zu berechnen: as _paasaa in views or as virtual field in the project?

Then check if new project variables could be used:
![qgep_project_variables](https://github.com/QGEP/QGEP/assets/1265865/7d3816ed-95d0-4599-9999-901aa3713991)

- [ ] Translation-transifex: is there a way to reuse the actual translation-files 

 - [ ] In transifex, we should not translate every value_en with value_de, because then also the fieldname in the value-list will translate and then you have two fields with value_de and in the attributeform there are English values… Don’t know if a solution on db-side is possible or in transifex (filter all fieldnames of vl_tables?)?

**Comment by sjib** [comment]
I added some screenshots for illustration

**Reply by ponceta** [reply]
Thank you a lot!

**Comment by cymed** [comment]
>     * [ ]  In transifex, we should not translate every value_en with value_de, because then also the fieldname in the value-list will translate and then you have two fields with value_de and in the attributeform there are English values… Don’t know if a solution on db-side is possible or in transifex (filter all fieldnames of vl_tables?)?

Does the problem still exist when there is an alias set on value_en?

**Reply by ponceta** [reply]
Good observation, we need to give it a try!

**Comment by cymed** [comment]
>    * [ ]  (what are the correct constraints / defaults for identifier and obj_id-fields): obj_id:
>    -> four option in constraints checked, identifier: not set in database, but set defaultvalue = obj_id

I would personally like to set  defaultvalue = obj_id only for vw_qgep_wastewater_structure (and wastewater nodes respectively). Reach identifiers should be managed by the database, setting them manually is tedious and prone to errors. See [vsa wiki](https://vsa.ch/wiki/hinweise-zum-attribut-bezeichnung/). 

To set the reach identifiers, we will need a new function. I suggest the following mechanism:

![mechanism.svg](https://github.com/QGEP/QGEP/assets/102035587/29babb27-dabb-4722-8632-23262c23f9e1)
 
With reach-reach connections me must ensure to not enter an infinite loop

**Reply by ponceta** [reply]
I've set it as default : identifier_from - identifier_to but I faced some issues with reach-reach connexions leading to XXX - XXX - XXX

I would not put too much db intelligence into that since it's a bit user dependant. 

And once again the non accessibility of the vsa wiki is painful.

**Reply by cymed** [reply]
Quote from the wiki following:

Si possible, **il faut laisser au système de saisie l’attribution de la désignation de la conduite et non à la personne responsable (source d’erreurs)**

It then suggests three options for setting the identifier:

- identifier from_ws-to_ws
- identifier from_ws
- independent identifier

the first option is the most widespread, but has problems with parallel reaches.
The second one has problems with all wastewater structures with multiple outward reaches
The third one is the easiest to implement, but the most troublesome when mapping older TV protocols.
The above mentioned solution is similar to the first option, but a bit more robust

**Reply by cymed** [reply]
A suggestion for the labeling function:

```

-- FUNCTION: qgep_od.update_reach_identifier(text, boolean)
CREATE OR REPLACE FUNCTION qgep_od.update_reach_identifier(
	_obj_id text,
	_all boolean DEFAULT false)
    RETURNS void
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE PARALLEL UNSAFE
AS $BODY$
BEGIN
UPDATE qgep_od.wastewater_networkelement  
SET 
	identifier=foo.identifier
FROM (
	SELECT
	    re.obj_id, -- reach.obj_id=wastewater_networkelement.obj_id
        coalesce(qgep_od.select_reach_identifier_part_by_rp(fk_reach_point_from)
				 ||'-'
				 ||qgep_od.select_reach_identifier_part_by_rp(fk_reach_point_to),obj_id) 
				 as identifier
    FROM 
		qgep_od.reach re
	WHERE 
		_all 
	OR 
		re.obj_id = _obj_id
	)foo
WHERE 
	foo.obj_id=wastewater_networkelement.obj_id;
END
$BODY$;

ALTER FUNCTION qgep_od.update_reach_identifier(text, boolean)
    OWNER TO qgep_sysadmin;
	
-- FUNCTION: qgep_od.update_reach_identifier(text, boolean)
CREATE OR REPLACE FUNCTION qgep_od.select_reach_identifier_part_by_rp(
	_obj_id text)
    RETURNS text
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE PARALLEL UNSAFE
AS $BODY$
/*
This Function is used in automated definition of identifier names. 
Input: object id of rech point
output: substring used in setting ne.identifier on reaches  
*/
DECLARE
    identifier_new text;
BEGIN

With disambiguation as(
SELECT
    rp.obj_id,
	ne.identifier~'((?:ch){1}(?:[[:alnum:]]){14})' as is_obj_id,
	re.obj_id is NULL as is_node,
	ne_check.obj_id is NULL as unique_postfix,
	right(ne.obj_id,8) as obj_id_postfix,
    left(ne.identifier,9) as trim_identifier
	
FROM qgep_od.reach_point rp
		LEFT JOIN qgep_od.wastewater_networkelement ne ON rp.fk_wastewater_networkelement=ne.obj_id
		LEFT JOIN qgep_od.reach re on re.obj_id=ne.obj_id
		LEFT JOIN qgep_od.wastewater_networkelement ne_check ON ne_check.identifier=right(ne.obj_id,8)
WHERE rp.obj_id=_obj_id
)
SELECT 
	CASE 
		WHEN is_node AND NOT is_obj_id THEN trim_identifier
		WHEN unique_postfix THEN obj_id_postfix
		ELSE NULL
	END
INTO identifier_new 
FROM disambiguation
WHERE disambiguation.obj_id=_obj_id;

RETURN identifier_new;
END
$BODY$;

ALTER FUNCTION qgep_od.select_reach_identifier_part_by_rp(text)
    OWNER TO qgep_sysadmin;
```

**Comment by sjib** [comment]
Documentation see https://github.com/QGEP/QGEP/discussions/828
